sql/mysqld.cc

```diff
@@ -6071,38 +6105,198 @@ void handle_connections_sockets()
   int flags=0,retval;
   st_vio *vio_tmp;
   bool is_unix_sock;
-#ifdef HAVE_POLL
+#ifdef HAVE_SYSTEMD
+  int systemd_n, systemd_fd;
+  union sockaddr_union socketpeer;
+  socklen_t socketpeer_len;
+  char const *family;
+#ifndef HAVE_POLL
+  int  pfs_flags[MAX_LISTEN_SOCKETS];
+  // systemd wants these too
+  MYSQL_SOCKET  pfs_fds[MAX_LISTEN_SOCKETS]; // used for non-polling interfaces
+#endif
+#endif // HAVE_SYSTEMD
   int socket_count= 0;
-  struct pollfd fds[3]; // for ip_sock, unix_sock and extra_ip_sock
-  MYSQL_SOCKET  pfs_fds[3]; // for performance schema
-#define setup_fds(X)                    \
+  short fd_type[MAX_LISTEN_SOCKETS];
+#ifdef HAVE_POLL
+  struct pollfd fds[MAX_LISTEN_SOCKETS]; // for ip_sock, extra_ip_sock and unix_sock (could be any order for systemd)
```

Comment:
"in any order" means that there can be no "extra_ip_sock" for systemd
activated sockets, right?

the point of extra_ip_sock is to have a different thread scheduler on it,
like, pool-of-threads on the ip_sock and thread-per-connection on
extra_ip_sock, for example as a backup admin connection when pool-of-threads
is full.

but sockets are coming in undefined order from systemd, then you cannot run
different schedulers on different sockets.

options:
* accept it as a fact and document that different schedulers on different
  sockets won't work. drawback: minus one feature
* use sd_is_socket_inet to find what socket is ip_sock and what is
  extra_ip_sock.  drawback: ports must be defined both in the
  mariadb.socket file and in my.cnf.
* pick unix_sock and ip_sock from systemd, but always open extra_ip_sock
  explicitly in mysqld. drawback: no socket activation for extra_ip_sock.

I think I'd prefer the third approach, as extra_ip_sock is an obscure feature,
only used as a backup connection.